cmd/containerd-shim-runhcs-v1/service_internal.go

```diff
@@ -123,52 +116,10 @@ func (s *service) createInternal(ctx context.Context, req *task.CreateTaskReques
 		}
 	}
 
-	var layerFolders []string
-	if spec.Windows != nil {
-		layerFolders = spec.Windows.LayerFolders
-	}
-	if err := validateRootfsAndLayers(req.Rootfs, layerFolders); err != nil {
-		return nil, err
-	}
-
-	// Only work with Windows here.
-	// Parsing of the rootfs mount for Linux containers occurs later.
-	if spec.Linux == nil && len(req.Rootfs) > 0 {
-		// For Windows containers, we work with LayerFolders throughout
-		// much of the creation logic in the shim. If we were given a
-		// rootfs mount, convert it to LayerFolders here.
-		m := req.Rootfs[0]
-		if m.Type != LegacyMountType && m.Type != CimFSMountType {
-			return nil, fmt.Errorf("unsupported Windows mount type: %s", m.Type)
-		} else if m.Type == CimFSMountType && (shimOpts.SandboxIsolation == runhcsopts.Options_HYPERVISOR) {
-			// For CIMFS layers only process isolation is supported right now.
-			return nil, fmt.Errorf("cimfs doesn't support hyperv isolation")
-		}
-
-		source, parentLayerPaths, err := parseLegacyRootfsMount(m)
-		if err != nil {
-			return nil, err
-		}
-
-		// Append the parents
-		spec.Windows.LayerFolders = append(spec.Windows.LayerFolders, parentLayerPaths...)
-		// Append the scratch
-		spec.Windows.LayerFolders = append(spec.Windows.LayerFolders, source)
-
-		if m.Type == CimFSMountType {
-			// write the layers to a file so that it can be used for proper cleanup during shim
-			// delete. We can't write to the config.json as it is read-only for shim.
-			f, err = os.Create(filepath.Join(req.Bundle, layersFile))
-			if err != nil {
-				return nil, err
-			}
-			if err := json.NewEncoder(f).Encode(spec.Windows.LayerFolders); err != nil {
-				f.Close()
-				return nil, err
-			}
-			f.Close()
-		}
-	}
+	// TODO(ambarve): With new changes in containerd, `shim delete` will be called
+	// only once per sandbox. That means we need to maintain information about each
+	// container in sandbox's storage and then ensure we cleanup all of those during
+	// `shim delete` in case the shim has crashed.
```

Comment:
Little confused why there is a TODO here? Will changes be needed to this part of the code to support cleanup?

Reply:
Earlier we had code here that stored the layer information (serialized rootfs or LayerFolders array) in a json file and that was used later during the `shim delete` call. That code was removed and this TODO was added here. This TODO is basically a reminder for us to figure out a proper way for storing layers of every container that runs under this shim and then properly clean all of them up when `shim delete` is called.

Reply:
I guess I'm wondering if we need both this TODO and the one in delete.go, or if we can just keep the delete.go one (especially since we don't actually parse layers right here anymore).

Reply:
Yeah, I only kept the TODO in delete.go and removed it from here.